include/utility/array.h

```diff
@@ -0,0 +1,197 @@
+/*************************
+ * @file array.h
+ * @author zhuzhile08 (zhuzhile08@gmail.com)
+ * 
+ * @brief a basic array implementation
+ * 
+ * @date 2022-12-24
+ * 
+ * @copyright Copyright (c) 2022
+ *************************/
+#pragma once
+
+#include <utility>
+#include <algorithm>
+
+#define NODISCARD [[nodiscard]]
+#define DEPRECATED [[deprecated]]
+
+namespace utility {
+
+// basic C-array wrapper implementation
+template <class Ty, size_t Size> struct Array {
```

Comment:
Why is this needed when `std::array` exists and is in the PCH already?

Reply:
Sorry, bad habit of mine, I like to use my own containers with a simpler and more readable implementation, which also means shorter headers, if you think I should still add this, I'd gladly redo some stuff here.

Well, not too sure about thy array is still in the PCHs, if again, if I should still include this, then I'd also create a PR to replace some PCHs

I guess that the same also goes for the smart pointer and the function pointer, but I have once benchmarked my smart pointer implementation to be just a tiny bit fasters than STD's and the function pointer should also perform better overall, again, because of the much simpler implementation. This one I haven't benchmarked though.